src/RequireCallStack.hs

```diff
@@ -48,9 +52,43 @@ import RequireCallStack.Internal
 -- @since 0.1.0.0
 type RequireCallStack = (HasCallStack, RequireCallStackImpl)
 
--- | Raise an 'ErrorCall' and incur a 'RequireCallStack' constraint while
--- you do so. This
+-- | If you're running into this class, then you need to add
+-- 'RequireCallStack' to your function's signature, or discharge the
+-- constraint using 'provideCallStack'.
+--
+-- @since 0.1.0.0
+type RequireCallStackImpl = ?provideCallStack :: CallStack
+
+-- | The constructor for this type is intentionally not exported
+data CallStack = CallStack
```

Comment:
nit: 
would you consider naming it slightly different such that won't be confused with the `GHC.Stack.CallStack`. 
`Confused` means when looking at code `?provideCallStack :: CallStack`, the first impression is the `CallStack` type is `GHC.Stack.CallStack` but later realized it is this `CallStack`.